// ebay_api/response.js
var _ = require('underscore');

/*
 * GET ORDERS RESPONSE
 */
 var OrdersResponse = function(response) {
        this.response = response;
        this.totalOrders = parseInt(this.response.PaginationResult.TotalNumberOfEntries);
        this.TimeStamp = this.response.TimeStamp;
        this._processOrders();
        this.allOrders = this.singleLineOrders.concat(this.multipleLineOrders);
        var ordersByID = {};
        this.allOrders.forEach(function(o) {
            ordersByID[o._id] = o;
        });
        this.ordersByID = ordersByID;
        this._getOrderLineItemIDs();
    }

OrdersResponse.prototype._getOrderLineItemIDs = function() {
    this.orderLineItemIDs = []
    var self = this;
    this.multipleLineOrders.forEach(function(o) {
        o.TransactionArray.Transaction.forEach(function(t) {
            
      self.orderLineItemIDs.push(t.OrderLineItemID);  
        });
    })

}

OrdersResponse.prototype._processOrders = function() {
    var self = this;
    this.multipleLineOrders = [];
    this.singleLineOrders = []
    var orders = this.response.OrderArray.Order;
    if (orders.length) {
        orders.forEach(function(o) {
            o._id = o.OrderID;
            if (o.OrderID.match(/-/)) {
                self.singleLineOrders.push(o);
            }
            else {
                self.multipleLineOrders.push(o);
            }
        });
    }
    else {
        orders._id = orders.OrderID;
        if(orders.OrderID.match(/-/)) {
            self.singleLineOrders.push(orders);
        }
        else {
            self.multipleLineOrders.push(orders);
        }            
    }
};

module.exports = OrdersResponse;
