// import_orders/ebay_getorders_factory.js
var Factory = require('factory');

Factory.define("OrderResponse", {
    TimeStamp: '2012-04-03T11:55:31.499Z',
    Ack: 'Success',
    PaginationResult: {
        TotalNumberOfPages: '1',
        TotalNumberOfEntries: '4'
    },
    HasMoreOrders: 'false',
    OrderArray: {
        Order: null
    },
    OrdersPerPage: '100',
    PageNumber: '1',
    ReturnedOrderCountActual: '4'
});

Factory.define("Order", {
    OrderID: 1,
    OrderStatus: 'Completed',
    AdjustmentAmount: {
        '#': '0.0',
        '@': []
    },
    AmountPaid: {
        '#': '6.0',
        '@': []
    },
    CheckoutStatus: {
        LastModifiedTime: '2012-04-02T10:32:31.000Z',
        PaymentMethod: 'PersonalCheck'
    },
    ShippingDetails: null,
    CreatedTime: '2012-04-02T10:32:31.000Z',
    Total: {
        '#': '6.0',
        '@': []
    },
    BuyerUserID: 'dizzynewmedia',
    EIASToken: 'nY+sHZ2PrBmdj6wVnY+sEZ2PrA2dj6wFk4CoAZeCoAqdj6x9nY+seQ==',
    TransactionArray: {
        Transaction: null
    }
});

Factory.define("Transaction", {
        Buyer: {
            Email: "david.pettifer@dizzy.co.uk"
        },
        ShippingDetails: {
            SellingManagerSalesRecordNumber: '148'
        },
        CreatedDate: '2012-04-02T10:40:29.000Z',
        Item: {
            ItemID: '110099445588',
            Site: 'UK',
            ConditionID: '1000',
            ConditionDisplayName: 'New'
        },
        Variation: {
            SKU: "25_NAME_SIMPLICITY_LINEN"  
        },
        QuantityPurchased: '5',
        Status: {
            PaymentHoldStatus: 'None'
        },
        TransactionID: '999888777',
        OrderLineItemID: '110099445588-999888777',
    });
    // Single Item 
    // orderID: ItemID-transactionID
    // transactionID: random generated number
    // Order line Item ID: same as orderID (ItemID-transactionID)
    
    // Multiple Item
    // orderID: random generated number
    // transactionID: random generated number
    // order line item ID: ItemID + transactionID

module.exports = Factory;
